app/services/mentee_application_transition_service.rb

```diff
@@ -0,0 +1,72 @@
+module MenteeApplicationTransitionService
+  class InvalidTransitionError < StandardError; end
+  extend self
+
+  STATUS_TRANSITION_MAPPING = {
+    application_received: {
+      valid_transitions: %i[promote reject],
+      promote_transition: :coding_challenge_sent
+    },
+    coding_challenge_sent: {
+      valid_transitions: [:promote],
+      promote_transition: :coding_challenge_received
+    },
+    coding_challenge_received: {
+      valid_transitions: %i[promote reject],
+      promote_transition: :coding_challenge_approved
+    },
+    coding_challenge_approved: {
+      valid_transitions: [:promote],
+      promote_transition: :phone_screen_scheduled
+    },
+    phone_screen_scheduled: {
+      valid_transitions: [:promote],
+      promote_transition: :phone_screen_completed
+    },
+    phone_screen_completed: {
+      valid_transitions: %i[promote reject],
+      promote_transition: :accepted
+    },
+    accepted: {
+      valid_transitions: []
+    },
+    rejected: {
+      valid_transitions: []
+    }
+  }.freeze
+
+  def promote!(application:, user:, note: nil)
```

Comment:
We should rename this `user` to something more clear (I got confused it was the person who is accepting/rejecting). Idk of a good name for them but it shouldn't be just user.